web/src/test/java/com/epam/rft/atsy/web/controllers/NewApplicationControllerTest.java

```diff
@@ -1,152 +1,248 @@
+  @Mock
+  private PositionService positionService;
+  @Mock
+  private ChannelService channelService;
+  @Mock
+  private MessageKeyResolver messageKeyResolver;
+  @InjectMocks
+  private NewApplicationController newApplicationController;
+
+  private ChannelDTO
+      channelDTO =
+      ChannelDTO.builder().id(1L).name(CHANNEL_NAME_FACEBOOK).deleted(false).build();
+
+  private PositionDTO
+      positionDTO =
+      PositionDTO.builder().id(1L).name(POSITION_NAME_DEVELOPER).deleted(false).build();
+  private StateDTO stateDTO = StateDTO.builder().id(1L).name(NEW_STATE).build();
+
+  private StateHistoryDTO stateHistoryDTO =
+      StateHistoryDTO.builder().candidateId(1L).channel(channelDTO).position(positionDTO)
+          .stateDTO(stateDTO).description(DESCRIPTION).build();
+
+  @Override
+  protected Object[] controllersUnderTest() {
+    return new Object[]{newApplicationController};
+  }
+
+  @Override
+  protected void addBeans(List<Object> beans) {
+    beans.add(this.candidateService);
+    beans.add(this.positionService);
+    beans.add(this.channelService);
+  }
+
+  @Test
+  public void loadPageShouldRenderNewApplicationView() throws Exception {
+    this.mockMvc.perform(get(REQUEST_URL_GET))
+        .andExpect(status().isOk())
+        .andExpect(view().name(VIEW_NAME))
+        .andExpect(forwardedUrl(VIEW_PREFIX + VIEW_NAME + VIEW_SUFFIX));
+  }
+
+  @Test
+  public void saveOrUpdateShouldBeSuccessful() throws Exception {
+    given(this.candidateService.getCandidate(1L)).willReturn(new CandidateDTO());
+    given(this.positionService.getPositionDtoById(1L)).willReturn(positionDTO);
+    given(this.channelService.getChannelDtoById(1L)).willReturn(channelDTO);
+
+    this.mockMvc.perform(post(REQUEST_URL_POST)
+        .param("candidateId", STRING_VALUE_ONE)
+        .param("position.id", STRING_VALUE_ONE).param("position.name", POSITION_NAME_DEVELOPER)
+        .param("channel.id", STRING_VALUE_ONE).param("channel.name", CHANNEL_NAME_FACEBOOK)
+        .param("description", DESCRIPTION))
+        .andExpect(status().is3xxRedirection())
+        .andExpect(redirectedUrl(REDIRECT_URL_FOR_CANDIDATE_A));
+
+    val applicationDTOCaptor = ArgumentCaptor.forClass(ApplicationDTO.class);
+    val stateHistoryDTOCaptor = ArgumentCaptor.forClass(StateHistoryDTO.class);
+
+    verify(applicationsService).saveApplication(applicationDTOCaptor.capture(), stateHistoryDTOCaptor.capture());
+    then(applicationsService).should()
+        .saveApplication(applicationDTOCaptor.getValue(), stateHistoryDTOCaptor.getValue());
+    then(this.candidateService).should().getCandidate(1L);
+    then(this.positionService).should().getPositionDtoById(1L);
+    then(this.channelService).should().getChannelDtoById(1L);
+
+    assertThat(stateHistoryDTOCaptor.getValue(), equalTo(stateHistoryDTO));
+    assertApplicationDtoWhenSaveOrUpdateIsSuccess(applicationDTOCaptor);
+  }
+
+  @Test(expected = NestedServletException.class)

```

Comment:
The candidate validator should handle the assertion error from the `candidateService.getCandidate()`, so that the user should see a meaningful error message.

I think the assertion error can be handled the same way as when the candidate id is not found.